libavcodec/x86/vvc/vvc_dmvr.asm

```diff
@@ -0,0 +1,396 @@
+; /*
+; * Provide AVX2 luma dmvr functions for VVC decoding
+; * Copyright (c) 2024 Nuo Mi
+; *
+; * This file is part of FFmpeg.
+; *
+; * FFmpeg is free software; you can redistribute it and/or
+; * modify it under the terms of the GNU Lesser General Public
+; * License as published by the Free Software Foundation; either
+; * version 2.1 of the License, or (at your option) any later version.
+; *
+; * FFmpeg is distributed in the hope that it will be useful,
+; * but WITHOUT ANY WARRANTY; without even the implied warranty of
+; * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+; * Lesser General Public License for more details.
+; *
+; * You should have received a copy of the GNU Lesser General Public
+; * License along with FFmpeg; if not, write to the Free Software
+; * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+; */
+%include "libavutil/x86/x86util.asm"
+
+%define MAX_PB_SIZE             128
+
+SECTION_RODATA 32
+
+shift_12    dw 1 << (15 - (12 - 10))
+shift3_8    dw 1 << (15 - (8 - 6))
+shift3_10   dw 1 << (15 - (10 - 6))
+shift3_12   dw 1 << (15 - (12 - 6))
+pw_16       dw 16
+
+%if ARCH_X86_64
+
+%if HAVE_AVX2_EXTERNAL
+
+SECTION .text
+
+%define pstride (bd / 10 + 1)
+
+; LOAD(dst, src)
+%macro LOAD_W16 2
+%if bd == 8
+    pmovzxbw %1, %2
+%else
+    movu %1, %2
+%endif
+%endmacro
+
+%macro SHIFT_W16 2
+%if bd == 8
+    psllw %1, (10 - bd)
+%elif bd == 10
+    ; nothing
+%else
+    pmulhrsw  %1, %2
+%endif
+%endmacro
+
+%macro SAVE_W16 2
+    movu %1, %2
+%endmacro
+
+; DMVR_4xW16(dst, dst_stride, dst_stride3, src, src_stride, src_stride3)
+%macro DMVR_4xW16 6
+    LOAD_W16               m0, [%4]
+    LOAD_W16               m1, [%4 + %5]
+    LOAD_W16               m2, [%4 + 2 * %5]
+    LOAD_W16               m3, [%4 + %6]
+
+    SHIFT_W16              m0, m4
+    SHIFT_W16              m1, m4
+    SHIFT_W16              m2, m4
+    SHIFT_W16              m3, m4
+
+    SAVE_W16    [%1]         , m0
+    SAVE_W16    [%1 + %2]    , m1
+    SAVE_W16    [%1 + 2 * %2], m2
+    SAVE_W16    [%1 + %3]    , m3
+%endmacro
+
+; buf += -stride * h + off
+; DMVR_OFFSET(buf, stride, off)
+%macro DMVR_OFFSET 3
+    mov                   id, hd
+    imul                  iq, %2
+    sub                   %1, iq
+    lea                   %1, [%1 + %3]
+%endmacro
+
+;void ff_vvc_dmvr_%1_avx2(int16_t *dst, const uint8_t *src, ptrdiff_t src_stride,
+;    int height, intptr_t mx, intptr_t my, int width);
+%macro DMVR_AVX2 1
+cglobal vvc_dmvr_%1, 4, 9, 5, dst, src, ss, h, ds, ds3, w, ss3, i
+%define bd %1
+    mov                  dsq, MAX_PB_SIZE * 2
+    lea                 ss3q, [ssq*3]
+    lea                 ds3q, [dsq*3]
+
+    mov                   wd, wm
+
+%if %1 > 10
+    vpbroadcastw          m4, [shift_%1]
```

Comment:
I think it's generally preferred to do

` vpbroadcastw          m4, [shift_%1]`

and then `shift_12: times 2 dw 1 << (15 - (12 - 10))`

Reply:
On intel architectures it's faster especially for Alderlake it's about 3x. It also uses at least two execution ports so it's possible that you get more performance depending on what other ports instructions near it need. For Zen 3/4 performance is identical.

Alder Lake P
broadcastd
Latency = 5/8
Throughput = 0.33

broadcastw
Lat = 6/9
Throughput = 1

Reply:
![image](https://github.com/user-attachments/assets/bb922bcd-4a3f-490f-8a6d-f6e34046ddc2)

Reply:
`vpbroadcastd` will have better latency than  `vpbroadcastw` on Skylake

Reply:
@stone-d-chen the page is cute, where can I find it?